README.md
# WAV

This is a crate for reading in and writing out wave files. It supports
uncompressed PCM bit depths of 8, 16, 24 bits, and 32bit IEEE Float formats,
both with any number of channels. Unfortunately other types of data format
(e.g. compressed WAVE files) are not supported. There is also no support for
any metadata chunks or any chunks other than the `"fmt "` and `"data"` chunks.

## Example

```rust
use std::fs::File;
use std::path::Path;

let mut inp_file = File::open(Path::new("data/sine.wav"))?;
let (header, data) = wav::read(&mut inp_file)?;

let mut out_file = File::create(Path::new("data/output.wav"))?;
wav::write(header, &data, &mut out_file)?;
```


CHANGELOG.md
# WAV

## Version 1.0.0

* [#13](https://github.com/Fluhzar/WAV/issues/13) - Changed tests to not rely on disk I/O.
* [#11](https://github.com/Fluhzar/WAV/pull/11) - Added benchmarks.

## Version 0.6.0

* Added benchmark tests under the `benches` folder [faddde7](https://github.com/Fluhzar/WAV/commit/faddde76cece1be530653d9e305fcad798627d54)
* Added support for 32-bit IEEE floating point data types [1117830](https://github.com/Fluhzar/WAV/commit/111783032030900464ead46be5c8924cac67e114)

## Version 0.5.0

* [#4](https://github.com/Fluhzar/WAV/issues/4) -Modified the signature of `wav::write` to borrow the `BitDepth` parameter rather than own.
* [#6](https://github.com/Fluhzar/WAV/issues/6) - Updated `riff` dependency to `^1.0`.
* Refactored code to be a bit more clear.
* Fixed incorrect endian-ness of the 24-bit reading and writing.
* Changed various `Into` and `TryInto` impls to `From` and `TryFrom` respectively.
* Updated a `From` impl that could panic into a `TryFrom` to prevent said panic.

## Version 0.4.1

* Fixed issue of using index-access of chunk instead of iterator, causing issues if there are chunks between "fmt " and "data"

## Version 0.4.0

* Changed `read_wav` to `read` and `write_wav` to `write`.
* Updated documentation.

## Version 0.3.0

* Change `read_wav` and `write_wav` to accept generic IO types.
* Changed an error message.

## Version 0.2.0

* Renamed `read_file` to `read_wav` for consistency and clarity.

## Version 0.1.1

* Gives proper error when audio data that isn't in uncompressed PCM format.
* Ensures it's finds the correct chunks for header and data info, ignoring any extra metadata chunks.
* Minor documentation updates.


tests/test.rs
use std::io;

macro_rules! impl_wav_test {
    ( $(
        $F:ident {
            File: $f:expr,
            AudioFormat: $af:expr,
            ChannelCount: $cc:expr,
            SamplingRate: $sr:expr,
            BitsPerSample: $bi_p_sa:expr,
            BytesPerSample: $by_p_sa:expr,
            BytesPerSecond: $by_p_se:expr,
            IsBitsFn: $ibf:path
            $(,)*
        }
    ),* $(,)* ) => { $(
        #[test]
        fn $F() {
            let raw: &[u8] = include_bytes!($f);
            let mut inp: io::Cursor<Vec<u8>> = io::Cursor::new(raw.into());
            let mut out: io::Cursor<Vec<u8>> = io::Cursor::new(vec![0u8 ; raw.len()]);

            let (header, data) = wav::read(&mut inp).unwrap();

            assert_eq!(header.audio_format, $af);
            assert_eq!(header.channel_count, $cc);
            assert_eq!(header.sampling_rate, $sr);
            assert_eq!(header.bits_per_sample, $bi_p_sa);
            assert_eq!(header.bytes_per_sample, $by_p_sa);
            assert_eq!(header.bytes_per_second, $by_p_se);
            assert!($ibf(&data));

            wav::write(header, &data, &mut out).unwrap();

            for (i, o) in inp.into_inner().drain(0..).zip(out.into_inner().drain(0..)) {
                assert_eq!(i, o);
            }
        }
    )* };
}

impl_wav_test! {
    sine_8bit_48khz {
        File: "../data/sine_8bit_48khz.wav",
        AudioFormat: 0x01,
        ChannelCount: 2,
        SamplingRate: 48_000,
        BitsPerSample: 8,
        BytesPerSample: 2,
        BytesPerSecond: 96_000,
        IsBitsFn: wav::bit_depth::BitDepth::is_eight,
    },
    sine_16bit_48khz {
        File: "../data/sine_16bit_48khz.wav",
        AudioFormat: 0x01,
        ChannelCount: 2,
        SamplingRate: 48_000,
        BitsPerSample: 16,
        BytesPerSample: 4,
        BytesPerSecond: 192_000,
        IsBitsFn: wav::bit_depth::BitDepth::is_sixteen,
    },
    sine_24bit_48khz {
        File: "../data/sine_24bit_48khz.wav",
        AudioFormat: 0x01,
        ChannelCount: 2,
        SamplingRate: 48_000,
        BitsPerSample: 24,
        BytesPerSample: 6,
        BytesPerSecond: 288_000,
        IsBitsFn: wav::bit_depth::BitDepth::is_twenty_four,
    },
    sine_32bit_float_48khz {
        File: "../data/sine_32bit_float_48khz.wav",
        AudioFormat: 0x03,
        ChannelCount: 2,
        SamplingRate: 48_000,
        BitsPerSample: 32,
        BytesPerSample: 8,
        BytesPerSecond: 384_000,
        IsBitsFn: wav::bit_depth::BitDepth::is_thirty_two_float,
    },
}


src/header.rs
//! Contains items responsible for dealing with the `"fmt "` chunk of wave files.

use std::convert::TryFrom;

/// Value signifying PCM data.
pub const WAV_FORMAT_PCM: u16 = 0x01;
/// Value signifying IEEE float data.
pub const WAV_FORMAT_IEEE_FLOAT: u16 = 0x03;

/// Structure for the `"fmt "` chunk of wave files, specifying key information about the enclosed
/// data.
///
/// This struct supports only PCM and IEEE float data, which is to say there is no extra members for
/// compressed format data.
#[derive(Debug, Default, Copy, Clone, Hash, PartialEq, Eq)]
#[allow(missing_docs)]
pub struct Header {
    pub audio_format: u16,
    pub channel_count: u16,
    pub sampling_rate: u32,
    pub bytes_per_second: u32,
    pub bytes_per_sample: u16,
    pub bits_per_sample: u16,
}

impl Header {
    /// Creates a new Header object.
    ///
    /// ## Note
    ///
    /// While the [`crate::read`] and [`crate::write`] functions only support uncompressed PCM/IEEE
    /// for the audio format, the option is given here to select any audio format for custom
    /// implementations of wave features.
    ///
    /// ## Parameters
    ///
    /// * `audio_format` - Audio format. Only [`WAV_FORMAT_PCM`] (0x01) and
    ///                    [`WAV_FORMAT_IEEE_FLOAT`] (0x03) are supported.
    /// * `channel_count` - Channel count. The number of channels each sample has. Generally 1
    ///                     (mono) or 2 (stereo).
    /// * `sampling_rate` - Sampling rate (e.g. 44.1kHz, 48kHz, 96kHz, etc.).
    /// * `bits_per_sample` - Number of bits in each (sub-channel) sample. Generally 8, 16, 24, or
    ///                       32.
    ///
    /// ## Example
    ///
    /// ```
    /// let h = wav::Header::new(wav::header::WAV_FORMAT_PCM, 2, 48_000, 16);
    /// ```
    #[must_use]
    pub fn new(
        audio_format: u16,
        channel_count: u16,
        sampling_rate: u32,
        bits_per_sample: u16,
    ) -> Header {
        Header {
            audio_format,
            channel_count,
            sampling_rate,
            bits_per_sample,
            bytes_per_second: u32::from((bits_per_sample >> 3) * channel_count) * sampling_rate,
            bytes_per_sample: (bits_per_sample >> 3) * channel_count,
        }
    }
}

impl From<Header> for [u8; 16] {
    #[allow(clippy::shadow_unrelated)]
    fn from(h: Header) -> Self {
        let mut v: [u8; 16] = [0; 16];

        let b = h.audio_format.to_le_bytes();
        v[0] = b[0];
        v[1] = b[1];
        let b = h.channel_count.to_le_bytes();
        v[2] = b[0];
        v[3] = b[1];
        let b = h.sampling_rate.to_le_bytes();
        v[4] = b[0];
        v[5] = b[1];
        v[6] = b[2];
        v[7] = b[3];
        let b = h.bytes_per_second.to_le_bytes();
        v[8] = b[0];
        v[9] = b[1];
        v[10] = b[2];
        v[11] = b[3];
        let b = h.bytes_per_sample.to_le_bytes();
        v[12] = b[0];
        v[13] = b[1];
        let b = h.bits_per_sample.to_le_bytes();
        v[14] = b[0];
        v[15] = b[1];

        v
    }
}

impl From<[u8; 16]> for Header {
    fn from(v: [u8; 16]) -> Self {
        let audio_format = u16::from_le_bytes([v[0], v[1]]);
        let channel_count = u16::from_le_bytes([v[2], v[3]]);
        let sampling_rate = u32::from_le_bytes([v[4], v[5], v[6], v[7]]);
        let bytes_per_second = u32::from_le_bytes([v[8], v[9], v[10], v[11]]);
        let bytes_per_sample = u16::from_le_bytes([v[12], v[13]]);
        let bits_per_sample = u16::from_le_bytes([v[14], v[15]]);

        Header {
            audio_format,
            channel_count,
            sampling_rate,
            bytes_per_second,
            bytes_per_sample,
            bits_per_sample,
        }
    }
}

impl TryFrom<&[u8]> for Header {
    type Error = &'static str;

    /// ## Errors
    ///
    /// This function will return an error if the given slice is smaller than 16 bytes.
    fn try_from(v: &[u8]) -> Result<Self, Self::Error> {
        if v.len() < 16 {
            Err("Slice is smaller than the minimum-required 16 bytes")
        } else {
            let mut a: [u8; 16] = [0; 16];
            a.copy_from_slice(&v[0..16]);
            Ok(Header::from(a))
        }
    }
}


benches/read.rs
#![feature(test)]
use std::fs::File;
use std::io::Cursor;
use std::io::Read;
use wav::read;

extern crate test;
use test::{black_box, Bencher};

fn read_bench(b: &mut Bencher, data: &[u8]) {
    b.iter(|| {
        black_box(read(&mut Cursor::new(data))).unwrap();
    })
}

macro_rules! create_read_bench {
    ($name: ident , $path: literal) => {
        #[bench]
        fn $name(b: &mut Bencher) {
            let mut data = Vec::new();
            File::open($path).unwrap().read_to_end(&mut data).unwrap();
            read_bench(b, data.as_slice());
        }
    };
}

create_read_bench!(read_sine, "data/sine.wav");
create_read_bench!(read_sine_8bit_48khz, "data/sine_8bit_48khz.wav");
create_read_bench!(read_sine_16bit_48khz, "data/sine_16bit_48khz.wav");
create_read_bench!(read_sine_24bit_48khz, "data/sine_24bit_48khz.wav");
create_read_bench!(read_sine_32bit_float_48khz, "data/sine_32bit_float_48khz.wav");

/*#[bench]
fn read_sine(b: &mut Bencher){
    let mut data = Vec::new();
    File::open(SINE_PATH).unwrap().read_to_end(&mut data).unwrap();
    read_bench(b , data.as_slice());
}

#[bench]
fn read_sine_8_48(b: &mut Bencher){
    let mut data = Vec::new();
    File::open(SINE_8_48_PATH).unwrap().read_to_end(&mut data).unwrap();
    b.iter(||{
        black_box(read(&mut Cursor::new(&data))).unwrap();
    })
}
#[bench]
fn read_sine_8_48(b: &mut Bencher){
    let mut data = Vec::new();
    File::open(SINE_8_48_PATH).unwrap().read_to_end(&mut data).unwrap();
    b.iter(||{
        black_box(read(&mut Cursor::new(&data))).unwrap();
    })
}*/


benches/bench.rs
#![feature(test)]
use std::io::Cursor;
use std::io::Read;
use std::{cell::Cell, fs::File};
use wav::{bit_depth::BitDepth, header::Header, read, write};

extern crate test;
use test::{black_box, Bencher};

fn read_bench(b: &mut Bencher, data: &[u8]) {
    b.iter(|| {
        black_box(read(&mut Cursor::new(data))).unwrap();
    })
}

// fn write_bench(b: &mut Bencher, data: Cell<&mut [u8]>, header: Header, bd: BitDepth) {
//     b.iter(|| {
//         black_box(write(header, &bd, &mut Cursor::new(data.take()))).unwrap();
//     })
// }

macro_rules! create_benches {
    ($({$read_name:ident, $write_name:ident, $path:literal $(,)*}),* $(,)*) => {$(
        #[bench]
        fn $read_name(b: &mut Bencher) {
            let mut inp = Vec::new();
            File::open($path).unwrap().read_to_end(&mut inp).unwrap();
            read_bench(b, &inp);
        }

        // #[bench]
        // fn $write_name(b: &mut Bencher) {
        //     let mut inp = Vec::new();
        //     File::open($path).unwrap().read_to_end(&mut inp).unwrap();
        //     let len = inp.len();
        //     let (header, track) = read(&mut Cursor::new(inp)).unwrap();
        //     let mut out = vec![0u8 ; len];
        //     write_bench(b, Cell::new(&mut out), header, track);
        // }
    )*};
}

create_benches! {
    {read_sine, write_sine, "data/sine.wav"},
    {read_sine_8bit_48khz, write_sine_8bit_48khz, "data/sine_8bit_48khz.wav"},
    {read_sine_16bit_48khz, write_sine_16bit_48khz, "data/sine_16bit_48khz.wav"},
    {read_sine_24it_48khz, write_sine_24bit_48khz, "data/sine_24bit_48khz.wav"},
    {read_sine_32bit_float_48khz, write_sine_32bit_float_48khz, "data/sine_32bit_float_48khz.wav"},
}


src/bit_depth.rs
//! Contains items for dealing with the `"data"` chunk of wave files.

use std::convert::TryFrom;

/// The supported bit-depths and containers for the samples at each depth.
#[derive(Debug, PartialEq, Clone)]
#[allow(missing_docs)]
pub enum BitDepth {
    Eight(Vec<u8>),
    Sixteen(Vec<i16>),
    TwentyFour(Vec<i32>),
    ThirtyTwoFloat(Vec<f32>),
    Empty,
}

impl BitDepth {
    /// Returns `true` if the bit depth is [`Self::Eight`].
    #[must_use]
    pub fn is_eight(&self) -> bool {
        matches!(self, Self::Eight(..))
    }

    /// Returns the inner value of the enum if it is [`Self::Eight`].
    #[must_use]
    pub fn as_eight(&self) -> Option<&Vec<u8>> {
        if let Self::Eight(v) = self {
            Some(v)
        } else {
            None
        }
    }

    /// Attempts to convert into the inner value.
    ///
    /// ## Errors
    ///
    /// Fails if enum isn't the [`Self::Eight`] variant.
    pub fn try_into_eight(self) -> Result<Vec<u8>, Self> {
        if let Self::Eight(v) = self {
            Ok(v)
        } else {
            Err(self)
        }
    }

    /// Returns `true` if the bit depth is [`Self::Sixteen`].
    #[must_use]
    pub fn is_sixteen(&self) -> bool {
        matches!(self, Self::Sixteen(..))
    }

    /// Returns the inner value of the enum if it is [`Self::Sixteen`].
    #[must_use]
    pub fn as_sixteen(&self) -> Option<&Vec<i16>> {
        if let Self::Sixteen(v) = self {
            Some(v)
        } else {
            None
        }
    }

    /// Attempts to convert into the inner value.
    ///
    /// ## Errors
    ///
    /// Fails if enum isn't the [`Self::Sixteen`] variant.
    pub fn try_into_sixteen(self) -> Result<Vec<i16>, Self> {
        if let Self::Sixteen(v) = self {
            Ok(v)
        } else {
            Err(self)
        }
    }

    /// Returns `true` if the bit depth is [`Self::TwentyFour`].
    #[must_use]
    pub fn is_twenty_four(&self) -> bool {
        matches!(self, Self::TwentyFour(..))
    }

    /// Returns the inner value of the enum if it is [`Self::TwentyFour`].
    #[must_use]
    pub fn as_twenty_four(&self) -> Option<&Vec<i32>> {
        if let Self::TwentyFour(v) = self {
            Some(v)
        } else {
            None
        }
    }

    /// Attempts to convert into the inner value.
    ///
    /// ## Errors
    ///
    /// Fails if enum isn't the [`Self::TwentyFour`] variant.
    pub fn try_into_twenty_four(self) -> Result<Vec<i32>, Self> {
        if let Self::TwentyFour(v) = self {
            Ok(v)
        } else {
            Err(self)
        }
    }

    /// Returns `true` if the bit depth is [`Self::ThirtyTwoFloat`].
    #[must_use]
    pub fn is_thirty_two_float(&self) -> bool {
        matches!(self, Self::ThirtyTwoFloat(..))
    }

    /// Returns the inner value of the enum if it is [`Self::ThirtyTwoFloat`].
    #[must_use]
    pub fn as_thirty_two_float(&self) -> Option<&Vec<f32>> {
        if let Self::ThirtyTwoFloat(v) = self {
            Some(v)
        } else {
            None
        }
    }

    /// Attempts to convert into the inner value.
    ///
    /// ## Errors
    ///
    /// Fails if enum isn't the [`Self::ThirtyTwoFloat`] variant.
    pub fn try_into_thirty_two_float(self) -> Result<Vec<f32>, Self> {
        if let Self::ThirtyTwoFloat(v) = self {
            Ok(v)
        } else {
            Err(self)
        }
    }

    /// Returns `true` if the bit depth is [`Self::Empty`].
    #[must_use]
    pub fn is_empty(&self) -> bool {
        matches!(self, Self::Empty)
    }
}

impl Default for BitDepth {
    fn default() -> Self {
        BitDepth::Empty
    }
}

impl From<Vec<u8>> for BitDepth {
    fn from(v: Vec<u8>) -> Self {
        BitDepth::Eight(v)
    }
}

impl From<Vec<i16>> for BitDepth {
    fn from(v: Vec<i16>) -> Self {
        BitDepth::Sixteen(v)
    }
}

impl From<Vec<i32>> for BitDepth {
    fn from(v: Vec<i32>) -> Self {
        BitDepth::TwentyFour(v)
    }
}

impl From<Vec<f32>> for BitDepth {
    fn from(v: Vec<f32>) -> Self {
        BitDepth::ThirtyTwoFloat(v)
    }
}

impl TryFrom<BitDepth> for Vec<u8> {
    type Error = &'static str;

    /// ## Errors
    ///
    /// This function fails if `value` is not [`BitDepth::Eight`].
    fn try_from(value: BitDepth) -> Result<Self, Self::Error> {
        if let BitDepth::Eight(v) = value {
            Ok(v)
        } else {
            Err("Bit-depth is not 8")
        }
    }
}

impl TryFrom<BitDepth> for Vec<i16> {
    type Error = &'static str;

    /// ## Errors
    ///
    /// This function fails if `value` is not [`BitDepth::Sixteen`].
    fn try_from(value: BitDepth) -> Result<Self, Self::Error> {
        if let BitDepth::Sixteen(v) = value {
            Ok(v)
        } else {
            Err("Bit-depth is not 16")
        }
    }
}

impl TryFrom<BitDepth> for Vec<i32> {
    type Error = &'static str;

    /// ## Errors
    ///
    /// This function fails if `value` is not [`BitDepth::TwentyFour`].
    fn try_from(value: BitDepth) -> Result<Self, Self::Error> {
        if let BitDepth::TwentyFour(v) = value {
            Ok(v)
        } else {
            Err("Bit-depth is not 24")
        }
    }
}

impl TryFrom<BitDepth> for Vec<f32> {
    type Error = &'static str;

    /// ## Errors
    ///
    /// This function fails if `value` is not [`BitDepth::ThirtyTwoFloat`].
    fn try_from(value: BitDepth) -> Result<Self, Self::Error> {
        if let BitDepth::ThirtyTwoFloat(v) = value {
            Ok(v)
        } else {
            Err("Bit-depth is not 32bit float")
        }
    }
}


src/tuple_iterator.rs
pub(crate) struct PairIter<T> {
    a: Option<T>,
    b: Option<T>,
}

impl<T> PairIter<T> {
    pub(crate) fn new(pair: (T, T)) -> Self {
        Self {
            a: Some(pair.0),
            b: Some(pair.1),
        }
    }
}

impl<T> Iterator for PairIter<T>
where
    T: Copy,
{
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        match self.a {
            Some(a) => {
                self.a = None;
                Some(a)
            }
            None => match self.b {
                Some(b) => {
                    self.b = None;
                    Some(b)
                }
                None => None,
            },
        }
    }
}

pub(crate) struct TripletIter<T> {
    a: Option<T>,
    b: Option<T>,
    c: Option<T>,
}

impl<T> TripletIter<T> {
    pub(crate) fn new(triplet: (T, T, T)) -> Self {
        Self {
            a: Some(triplet.0),
            b: Some(triplet.1),
            c: Some(triplet.2),
        }
    }
}

impl<T> Iterator for TripletIter<T>
where
    T: Copy,
{
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        match self.a {
            Some(a) => {
                self.a = None;
                Some(a)
            }
            None => match self.b {
                Some(b) => {
                    self.b = None;
                    Some(b)
                }
                None => match self.c {
                    Some(c) => {
                        self.c = None;
                        Some(c)
                    }
                    None => None,
                },
            },
        }
    }
}

pub(crate) struct QuadrupletIter<T> {
    a: Option<T>,
    b: Option<T>,
    c: Option<T>,
    d: Option<T>,
}

impl<T> QuadrupletIter<T> {
    pub(crate) fn new(quadruplet: (T, T, T, T)) -> Self {
        Self {
            a: Some(quadruplet.0),
            b: Some(quadruplet.1),
            c: Some(quadruplet.2),
            d: Some(quadruplet.3),
        }
    }
}

impl<T> Iterator for QuadrupletIter<T>
where
    T: Copy,
{
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        match self.a {
            Some(a) => {
                self.a = None;
                Some(a)
            }
            None => match self.b {
                Some(b) => {
                    self.b = None;
                    Some(b)
                }
                None => match self.c {
                    Some(c) => {
                        self.c = None;
                        Some(c)
                    }
                    None => match self.d {
                        Some(d) => {
                            self.d = None;
                            Some(d)
                        }
                        None => None,
                    },
                },
            },
        }
    }
}


src/lib.rs
//! This is a crate for reading in and writing out wave files. It supports uncompressed PCM bit
//! depths of 8, 16, 24 bits, and 32bit IEEE Float formats, both with any number of channels.
//! Unfortunately other types of data format (e.g. compressed WAVE files) are not supported. There
//! is also no support for any metadata chunks or any chunks other than the `"fmt "` and `"data"`
//! chunks.
//!
//! ## Example
//!
//! ```rust
//! # fn main() -> std::io::Result<()> {
//! use std::fs::File;
//! use std::path::Path;
//!
//! let mut inp_file = File::open(Path::new("data/sine.wav"))?;
//! let (header, data) = wav::read(&mut inp_file)?;
//!
//! let mut out_file = File::create(Path::new("data/output.wav"))?;
//! wav::write(header, &data, &mut out_file)?;
//! # Ok(())
//! # }
//! ```

#![warn(missing_docs)]
#![warn(clippy::all)]
#![warn(clippy::pedantic)]

use std::{
    convert::TryFrom,
    io::{self, Read, Write},
};

pub mod header;
pub use header::{Header, WAV_FORMAT_IEEE_FLOAT, WAV_FORMAT_PCM};

pub mod bit_depth;
pub use bit_depth::BitDepth;

mod tuple_iterator;
use tuple_iterator::{PairIter, QuadrupletIter, TripletIter};

/// Reads in the given `reader` and attempts to extract the audio data and header from it.
///
/// ## Errors
///
/// This function fails under the following circumstances:
///
/// * Any error occurring from the `reader` parameter during reading.
/// * The data isn't RIFF data.
/// * The wave header specifies a compressed data format.
/// * The wave header specifies an unsupported bit-depth.
/// * The wave data is malformed, or otherwise couldn't be parsed into samples.
#[allow(clippy::similar_names)]
pub fn read<R>(reader: &mut R) -> io::Result<(Header, BitDepth)>
where
    R: Read + io::Seek,
{
    let header = read_header(reader)?;
    Ok((header, read_data(reader, &header)?))
}

/// Writes the given wav data to the given `writer`.
///
/// ## Notes
///
/// Although `track` is a borrowed value, its contents will be formatted into an owned `Vec<u8>` so
/// that it can be written to the `writer` through [`riff::ChunkContents::write`].
///
/// ## Errors
///
/// This function fails under the following circumstances:
///
/// * Any error occurring from the `writer` parameter during writing.
/// * The given [`BitDepth`] is [`BitDepth::Empty`].
pub fn write<W>(header: Header, track: &BitDepth, writer: &mut W) -> std::io::Result<()>
where
    W: Write + io::Seek,
{
    const WAVE_ID: riff::ChunkId = riff::ChunkId {
        value: [b'W', b'A', b'V', b'E'],
    };
    const HEADER_ID: riff::ChunkId = riff::ChunkId {
        value: [b'f', b'm', b't', b' '],
    };
    const DATA_ID: riff::ChunkId = riff::ChunkId {
        value: [b'd', b'a', b't', b'a'],
    };

    let h_vec: [u8; 16] = header.into();
    let h_dat = riff::ChunkContents::Data(HEADER_ID, Vec::from(h_vec));

    let d_vec = match track {
        BitDepth::Eight(v) => v.clone(),
        BitDepth::Sixteen(v) => v
            .iter()
            .flat_map(|s| {
                let v = s.to_le_bytes();
                PairIter::new((v[0], v[1]))
            })
            .collect::<Vec<_>>(),
        BitDepth::TwentyFour(v) => v
            .iter()
            .flat_map(|s| {
                let v = s.to_le_bytes().split_at(1).1.to_owned();
                TripletIter::new((v[0], v[1], v[2]))
            })
            .collect::<Vec<_>>(),
        BitDepth::ThirtyTwoFloat(v) => v
            .iter()
            .flat_map(|s| {
                let v = s.to_le_bytes().to_owned();
                QuadrupletIter::new((v[0], v[1], v[2], v[3]))
            })
            .collect::<Vec<_>>(),
        _ => {
            return Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                "Empty audio data given",
            ))
        }
    };
    let d_dat = riff::ChunkContents::Data(DATA_ID, d_vec);

    let r = riff::ChunkContents::Children(riff::RIFF_ID.clone(), WAVE_ID, vec![h_dat, d_dat]);

    r.write(writer)?;

    Ok(())
}

#[allow(clippy::similar_names)]
fn read_header<R>(reader: &mut R) -> io::Result<Header>
where
    R: Read + io::Seek,
{
    let wav = verify_wav_file(reader)?;

    for c in wav.iter(reader) {
        if c.id().as_str() == "fmt " {
            // Read header contents
            let header_bytes = c.read_contents(reader)?;
            let header = Header::try_from(header_bytes.as_slice())
                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;

            // Return error if not using PCM
            match header.audio_format {
                WAV_FORMAT_PCM | WAV_FORMAT_IEEE_FLOAT => return Ok(header),
                _ => {
                    return Err(io::Error::new(
                        io::ErrorKind::Other,
                        "Unsupported data format, data is not in uncompressed PCM format, aborting",
                    ))
                }
            };
        }
    }

    Err(io::Error::new(
        io::ErrorKind::InvalidData,
        "RIFF data is missing the \"fmt \" chunk, aborting",
    ))
}

#[allow(clippy::similar_names)]
fn read_data<R>(reader: &mut R, header: &Header) -> io::Result<BitDepth>
where
    R: Read + io::Seek,
{
    let wav = verify_wav_file(reader)?;

    for c in wav.iter(reader) {
        if c.id().as_str() == "data" {
            // Read data contents
            let data_bytes = c.read_contents(reader)?;

            let wav_data = match header.audio_format {
                WAV_FORMAT_PCM => match header.bits_per_sample {
                    8 => Ok(BitDepth::Eight(data_bytes)),
                    16 => Ok(BitDepth::Sixteen({
                        let mut tmpv = Vec::with_capacity(data_bytes.len() / 2);
                        tmpv.extend(
                            data_bytes
                                .chunks_exact(2)
                                .map(|i| i16::from_le_bytes([i[0], i[1]])),
                        );
                        tmpv
                    })),
                    24 => Ok(BitDepth::TwentyFour({
                        let mut tmpv = Vec::with_capacity(data_bytes.len() / 3);
                        tmpv.extend(
                            data_bytes
                                .chunks_exact(3)
                                .map(|i| i32::from_le_bytes([0, i[0], i[1], i[2]])),
                        );
                        tmpv
                    })),
                    _ => Err(io::Error::new(
                        io::ErrorKind::Other,
                        "Unsupported PCM bit depth",
                    )),
                },
                WAV_FORMAT_IEEE_FLOAT => match header.bits_per_sample {
                    32 => Ok(BitDepth::ThirtyTwoFloat({
                        let mut tmpv = Vec::with_capacity(data_bytes.len() / 4);
                        tmpv.extend(
                            data_bytes
                                .chunks_exact(4)
                                .map(|f| f32::from_le_bytes([f[0], f[1], f[2], f[3]])),
                        );
                        tmpv
                    })),
                    _ => Err(io::Error::new(
                        io::ErrorKind::Other,
                        "Unsupported IEEE Float bit depth",
                    )),
                },
                _ => Err(io::Error::new(
                    io::ErrorKind::Other,
                    "Unsupported WAV format",
                )),
            };

            return wav_data;
        }
    }

    Err(io::Error::new(
        io::ErrorKind::Other,
        "Could not parse audio data",
    ))
}

fn verify_wav_file<R>(reader: &mut R) -> io::Result<riff::Chunk>
where
    R: Read + io::Seek,
{
    let wav = riff::Chunk::read(reader, 0)?;

    let form_type = wav.read_type(reader)?;

    if form_type.as_str() == "WAVE" {
        Ok(wav)
    } else {
        Err(io::Error::new(
            io::ErrorKind::Other,
            "RIFF file type not \"WAVE\"",
        ))
    }
}
